src/azul/project/hca/transformers.py

```diff
@@ -224,3 +235,65 @@ def create_documents(self,
                                                                 version=bundle.version,
                                                                 contents=contents)])
             yield es_document
+
+
+class ProjectTransformer(Transformer):
+    @property
+    def entity_name(self):
+        return 'projects'
+
+    def create_documents(self,
+                         uuid: str,
+                         version: str,
+                         manifest: List[JSON],
+                         metadata_files: Mapping[str, JSON]
+                         ) -> Sequence[ElasticSearchDocument]:
+        # Create a bundle object.
+        bundle = api.Bundle(uuid=uuid,
+                            version=version,
+                            manifest=manifest,
+                            metadata_files=metadata_files)
+
+        bundle_uuid = str(bundle.uuid)
+        simplified_project = _project_dict(bundle)
+
+        # Create ElasticSearch documents
+        for project in bundle.projects.values():
+            contents = dict(specimens=self._get_unique_list_of_specimens(bundle),
+                            files=self._get_unique_list_of_files(bundle),
+                            processes=self._get_unique_list_of_process(bundle),
+                            project=simplified_project)
```

Comment:
I am concerned that this implementation of `ProjectTransformer` departs from existing patterns and introduces code duplication, which could lead to inconsistencies in the future.

Instead, I would try implementing `ProjectTransformer` something like this:
```
class ProjectTransformer(Transformer):
    @property
    def entity_name(self):
        return 'projects'

    def create_documents(self,
                         uuid: str,
                         version: str,
                         manifest: List[JSON],
                         metadata_files: Mapping[str, JSON]
                         ) -> Sequence[ElasticSearchDocument]:
        # Create a bundle object.
        bundle = api.Bundle(uuid=uuid,
                            version=version,
                            manifest=manifest,
                            metadata_files=metadata_files)

        # Create ElasticSearch documents

        # Gather information on files in the bundle
        for file in bundle.files.values():
            files_visitor = TransformerVisitor()

            # Visit the relatives
            file.accept(files_visitor)  # Visit descendants
            file.ancestors(files_visitor)

        # Gather information on specimens in the bundle
        for specimen in bundle.specimens:
            specimen_visitor = TransformerVisitor()

            # Visit the relatives
            specimen.accept(specimen_visitor)  # Visit descendants
            specimen.ancestors(specimen_visitor)

        # Assign the contents to the ES doc
        contents = dict(specimens=_specimen_dict(specimen_visitor.specimens),
                        files=list(files_visitor.files.values()),
                        processes=list(specimen_visitor.processes.values()),
                        project=_project_dict(bundle))
        es_document = ElasticSearchDocument(entity_type=self.entity_name,
                                            entity_id=str(project.document_id),
                                            bundles=[Bundle(uuid=str(bundle.uuid),
                                                            version=bundle.version,
                                                            contents=contents)])
        yield es_document
```

I think this is close to something that would work, although, to be clear, I have not yet tested it as I still need to look at how the test infrastructure and tests are setup for this stuff. Yet, you get the idea.

Reply:
As I mentioned in Slack, I agree with your suggestion and will work on the update.

Reply:
Thanks, @shiroyuki, I reposted this from our DM Slack chat so @hannes-ucsc has visibility to the comment history.

Also, on closer examination, I think the above proposal can be implemented using a single instance of `TransformerVisitor`. Please let me know if you have questions. Thanks!